File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> res(nums.size(),1);
        
        int currLeft = 1;
        for(int i=0;i<nums.size();i++){
            res[i] = currLeft;
            currLeft *= nums[i];
        }
        
        int currRight = 1;
        for(int i=nums.size()-1;i>=0;i--){
            res[i] *= currRight;
            currRight *= nums[i];
        }
        
        return res;
    }
};

File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans(n);
        
        //storing the prefix
        int prefix = 1;
        for(int i=0;i<n;i++) {
            ans[i] = prefix;
            prefix *= nums[i];
        }
        
        //storing the postfix from end to start.
        int postfix = 1;
        for(int i=n-1;i>=0;i--) {
            ans[i] *= postfix;
            postfix *= nums[i];
        }
        
        return ans;
    }
};